service/controller/v3/resource/instance/create.go

```diff
@@ -61,71 +49,75 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
-			// fall through
-		} else if err != nil {
-			return microerror.Mask(err)
-		} else {
-			p, err := key.ToMap(d.Properties.Parameters)
+		var k8sClient kubernetes.Interface
+		{
+			i := key.ClusterID(customObject)
+			e := key.ClusterAPIEndpoint(customObject)
+			k8sClient, err = r.guestCluster.NewK8sClient(ctx, i, e)
 			if err != nil {
 				return microerror.Mask(err)
 			}
-			v, ok := p[versionsKey]
-			if !ok {
-				// fall through
-			} else {
-				m, err := key.ToMap(v)
-				if err != nil {
-					return microerror.Mask(err)
-				}
-				v, err := key.ToKeyValue(m)
-				if err != nil {
-					return microerror.Mask(err)
-				}
-				masterVersionsValue, err = key.ToStringMap(v)
-				if err != nil {
-					return microerror.Mask(err)
-				}
-			}
 		}
-	}
 
-	var workerVersionsValue map[string]string
-	{
-		d, err := deploymentsClient.Get(ctx, key.ClusterID(customObject), "worker-vmss-deploy")
-		if IsDeploymentNotFound(err) {
+		o := metav1.ListOptions{}
+		list, err := k8sClient.CoreV1().Nodes().List(o)
+		if guest.IsAPINotAvailable(err) {
 			// fall through
+			r.logger.LogCtx(ctx, "level", "debug", "message", "not fetching node versions")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "guest API is not available")
 		} else if err != nil {
 			return microerror.Mask(err)
 		} else {
-			p, err := key.ToMap(d.Properties.Parameters)
-			if err != nil {
-				return microerror.Mask(err)
-			}
-			v, ok := p[versionsKey]
-			if !ok {
-				// fall through
-			} else {
-				m, err := key.ToMap(v)
-				if err != nil {
-					return microerror.Mask(err)
+			for _, node := range list.Items {
+				l := node.GetLabels()
+				n := node.GetName()
+
+				labelProvider := "giantswarm.io/provider"
+				p, ok := l[labelProvider]
+				if !ok {
+					return microerror.Maskf(missingLabelError, labelProvider)
 				}
-				v, err := key.ToKeyValue(m)
-				if err != nil {
-					return microerror.Mask(err)
+				labelVersion := p + "-operator.giantswarm.io/version"
+				v, ok := l[labelVersion]
+				if !ok {
+					return microerror.Maskf(missingLabelError, labelProvider)
 				}
-				workerVersionsValue, err = key.ToStringMap(v)
-				if err != nil {
-					return microerror.Mask(err)
+
+				role, ok := l["role"]
+				if ok && role == "master" {
+					masterVersionsValue[n] = v
+				} else {
+					workerVersionsValue[n] = v
 				}
 			}
 		}
 	}
 
+	var nodeConfigs []corev1alpha1.DrainerConfig
+	{
+		n := v1.NamespaceAll
+		o := metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s=%s", key.ClusterIDLabel, key.ClusterID(customObject)),
+		}
+
+		list, err := r.g8sClient.CoreV1alpha1().DrainerConfigs(n).List(o)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		nodeConfigs = list.Items
+	}
+
 	var allMasterInstances []compute.VirtualMachineScaleSetVM
+	var drainedMasterInstance *compute.VirtualMachineScaleSetVM
 	var reimagedMasterInstance *compute.VirtualMachineScaleSetVM
 	var updatedMasterInstance *compute.VirtualMachineScaleSetVM
```

Comment:
There is no need to declare those vars outside the block. This gives notion they will be used further.

Also I think those names should be `masterInstanceToDrain` etc.

Same applies to worker.

Reply:
This derailed a bit from the prototyping face. This was all a bit different in the first place and now it is not needed that way anymore. Will provide a PR.